### ISSUE TITLE
enterprise autoscaling issues, indefinitely queued jobs within workflows

### ISSUE THREAD

**Issue by kathleenfrench** [actions-runner-controller/actions-runner-controller#470]
hey - first off, thanks for all of your work on this 😄  - it's a great project!

apologies in advance for the lengthy note -- just hoping to provide as much helpful info as possible upfront. nevertheless, a (poor) attempt at the cliffsnotes version:

> when a workflow with more jobs than currently-running runner pods is kicked off, _sometimes_ the only jobs that **ever** run - even if a scale up is later triggered (though that scale-up is never triggered within the `scaleUp` period) - are those that are initially picked up. _however_, sometimes the remaining queued jobs _do_ run, but only after the first jobs complete. there are no errors in the controller logs in either instance.

**scenario**
> 1 workflow with 4 jobs
> 10s `syncPeriod`

`minReplicas`: 2
`maxReplicas`: 20
`scaleUpThreshold`: 0.5
`scaleUpFactor`: 3.0

**expected behavior**
- 2 jobs immediately picked up by the 2 listening pods
- trigger a scale up after the 10s `syncPeriod` completes as > `0.5` of the running pods are now 'busy' 
- this will result in `6` runner pods
- the remaining 2 jobs in the workflow would get picked up by the newly created runner pods and run alongside the 2 already running

**actual behavior**
- 2 jobs immediately picked up by the 2 listening pods
- no scale up after 10s `syncPeriod` - controller logs show:
```
2021-04-21T18:12:14.019Z        DEBUG   actions-runner-controller.horizontalrunnerautoscaler    Calculated desired replicas of 2      {"horizontalrunnerautoscaler": "github-actions/ghe-runner-deployment-autoscaler", "suggested": 1, "reserved": 0, "min": 2, "cached": 1, "max": 20}
```
- once a job completes its pod terminates, a new pod is created, and a queued job runs. at this point, no scale up has been triggered yet
- scale up to 6 pods seems to coincide subsequent to the 3rd job starting on a newly created pod (when the total number of pods is still 2) **but** none of the remaining queued jobs seem to get picked up by these pods created as part of that scale up, but instead wait for _new_ pods using the same deployment from the initial jobs that did run

<details>
<summary>corresponding scale up logs (click me)</summary>

```
2021-04-21T18:16:34.025Z        DEBUG   actions-runner-controller.horizontalrunnerautoscaler    Calculated desired replicas of 2      {"horizontalrunnerautoscaler": "github-actions/ghe-runner-deployment-autoscaler", "suggested": 1, "reserved": 0, "min": 2, "cached": 1, "max": 20}
2021-04-21T18:16:34.025Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "horizontalrunnerautoscaler-controller", "request": "github-actions/ghe-runner-deployment-autoscaler"}
2021-04-21T18:16:34.025Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runnerreplicaset-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn"}
2021-04-21T18:16:34.025Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runnerdeployment-controller", "request": "github-actions/ghe-global-runner-deployment"}
2021-04-21T18:16:34.087Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runner-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn-phgmf"}
2021-04-21T18:16:34.147Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runner-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn-kxjgg"}
2021-04-21T18:16:34.985Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runner-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn-kxjgg"}
2021-04-21T18:16:35.047Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runner-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn-phgmf"}
2021-04-21T18:16:37.204Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runnerreplicaset-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn"}
2021-04-21T18:16:37.204Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runnerdeployment-controller", "request": "github-actions/ghe-global-runner-deployment"}
2021-04-21T18:16:44.026Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runnerdeployment-controller", "request": "github-actions/ghe-global-runner-deployment"}
2021-04-21T18:16:44.026Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runnerreplicaset-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn"}
2021-04-21T18:16:44.026Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runnerreplicaset-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn"}
2021-04-21T18:16:44.096Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runner-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn-phgmf"}
2021-04-21T18:16:44.159Z        DEBUG   actions-runner-controller.horizontalrunnerautoscaler    Suggested desired replicas of 6 by PercentageRunnersBusy       {"replicas_desired_before": 2, "replicas_desired": 6, "num_runners": 2, "num_runners_registered": 2, "num_runners_busy": 2, "namespace": "github-actions", "runner_deployment": "ghe-global-runner-deployment", "horizontal_runner_autoscaler": "ghe-runner-deployment-autoscaler", "enterprise": "[redacted]", "organization": "", "repository": ""}
2021-04-21T18:16:44.159Z        DEBUG   actions-runner-controller.horizontalrunnerautoscaler    Calculated desired replicas of 6      {"horizontalrunnerautoscaler": "github-actions/ghe-runner-deployment-autoscaler", "suggested": 6, "reserved": 0, "min": 2, "max": 20}
2021-04-21T18:16:44.169Z        DEBUG   controller-runtime.webhook.webhooks     received request        {"webhook": "/mutate-actions-summerwind-dev-v1alpha1-runnerdeployment", "UID": "32d77506-8697-404f-a366-5804eecb6885", "kind": "actions.summerwind.dev/v1alpha1, Kind=RunnerDeployment", "resource": {"group":"actions.summerwind.dev","version":"v1alpha1","resource":"runnerdeployments"}}
2021-04-21T18:16:44.170Z        DEBUG   controller-runtime.webhook.webhooks     wrote response  {"webhook": "/mutate-actions-summerwind-dev-v1alpha1-runnerdeployment", "UID": "32d77506-8697-404f-a366-5804eecb6885", "allowed": true, "result": {}, "resultError": "got runtime.Object without object metadata: &Status{ListMeta:ListMeta{SelfLink:,ResourceVersion:,Continue:,RemainingItemCount:nil,},Status:,Message:,Reason:,Details:nil,Code:200,}"}
2021-04-21T18:16:44.178Z        DEBUG   controller-runtime.webhook.webhooks     received request        {"webhook": "/validate-actions-summerwind-dev-v1alpha1-runnerdeployment", "UID": "b5d15de0-63d7-40a3-8225-7b0a8fcb7fe6", "kind": "actions.summerwind.dev/v1alpha1, Kind=RunnerDeployment", "resource": {"group":"actions.summerwind.dev","version":"v1alpha1","resource":"runnerdeployments"}}
2021-04-21T18:16:44.179Z        INFO    runnerdeployment-resource       validate resource to be updated {"name": "ghe-global-runner-deployment"}
2021-04-21T18:16:44.179Z        DEBUG   controller-runtime.webhook.webhooks     wrote response  {"webhook": "/validate-actions-summerwind-dev-v1alpha1-runnerdeployment", "UID": "b5d15de0-63d7-40a3-8225-7b0a8fcb7fe6", "allowed": true, "result": {}, "resultError": "got runtime.Object without object metadata: &Status{ListMeta:ListMeta{SelfLink:,ResourceVersion:,Continue:,RemainingItemCount:nil,},Status:,Message:,Reason:,Details:nil,Code:200,}"}
2021-04-21T18:16:44.180Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "runner-controller", "request": "github-actions/ghe-global-runner-deployment-rwtwn-kxjgg"}
2021-04-21T18:16:44.194Z        DEBUG   controller-runtime.webhook.webhooks     received request        {"webhook": "/mutate-actions-summerwind-dev-v1alpha1-runnerreplicaset", "UID": "861c2f84-6aba-4e16-922e-84ae490b0190", "kind": "actions.summerwind.dev/v1alpha1, Kind=RunnerReplicaSet", "resource": {"group":"actions.summerwind.dev","version":"v1alpha1","resource":"runnerreplicasets"}}
2021-04-21T18:16:44.195Z        DEBUG   controller-runtime.webhook.webhooks     wrote response  {"webhook": "/mutate-actions-summerwind-dev-v1alpha1-runnerreplicaset", "UID": "861c2f84-6aba-4e16-922e-84ae490b0190", "allowed": true, "result": {}, "resultError": "got runtime.Object without object metadata: &Status{ListMeta:ListMeta{SelfLink:,ResourceVersion:,Continue:,RemainingItemCount:nil,},Status:,Message:,Reason:,Details:nil,Code:200,}"}
2021-04-21T18:16:44.196Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "horizontalrunnerautoscaler-controller", "request": "github-actions/ghe-runner-deployment-autoscaler"}
2021-04-21T18:16:44.197Z        DEBUG   actions-runner-controller.horizontalrunnerautoscaler    Calculated desired replicas of 6      {"horizontalrunnerautoscaler": "github-actions/ghe-runner-deployment-autoscaler", "suggested": 6, "reserved": 0, "min": 2, "cached": 6, "last_scale_up_time": "2021-04-21 18:16:44 +0000 UTC", "scale_down_delay_until": "2021-04-21 18:16:54 +0000 UTC", "max": 20}
```

</details>

_job that ran_
```
Current runner version: '2.277.1'
Runner name: 'ghe-global-runner-deployment-rwtwn-kxjgg'
Runner group name: 'k8s'
Machine name: 'ghe-global-runner-deployment-rwtwn-kxjgg'
```
_job that was queued until one of the initial jobs completed and its pod terminated_
```
Current runner version: '2.277.1'
Runner name: 'ghe-global-runner-deployment-rwtwn-kxjgg'
Runner group name: 'k8s'
Machine name: 'ghe-global-runner-deployment-rwtwn-kxjgg'
```

## overview

i've confirmed this for clean installs (via helm for the controller, having deleted and re-created the CRDs manually per existing issue instructions) using both `v0.18.2` as well as the `canary` actions-controller image.

_versioning_
```
Name:                   actions-runner-controller
Namespace:              actions-runner-system
CreationTimestamp:      Wed, 14 Apr 2021 17:28:40 -0500
Labels:                 app.kubernetes.io/instance=actions-runner-controller
                        app.kubernetes.io/managed-by=Helm
                        app.kubernetes.io/name=actions-runner-controller
                        app.kubernetes.io/version=0.18.2
                        helm.sh/chart=actions-runner-controller-0.11.0
Annotations:            deployment.kubernetes.io/revision: 17
                        meta.helm.sh/release-name: actions-runner-controller
                        meta.helm.sh/release-namespace: actions-runner-system
Selector:               app.kubernetes.io/instance=actions-runner-controller,app.kubernetes.io/name=actions-runner-controller
```

```
## controller
image:
  repository: summerwind/actions-runner-controller
   tag: "v0.18.2" (also w/ canary)
```

i've configured the controller on the enterprise level, as we have many organizations to support, though right now it's only enabled for 2 organizations and a select number of repositories within each. given we can disable rate-limiting, i haven't run into any issues keeping the `syncPeriod` low on that front.

so far, there's not been any issues when it comes to the runners themselves being recognized on the github side (the workers all appear in the custom `k8s` group created), and when the pod(s) for a job spin up, they select the appropriate runner deployment per the label in that spec that corresponds to what's referenced in the actions workflow itself.

the problem is it doesn't seem that the `PercentageRunnersBusy` metric i've used to autoscale is correctly registering the number of queued jobs. in debugging the controller side, i've confirmed there are no errors in the logs themselves. the output indicates that the desired number of replicas has been successfully reconciled; the problem is those values aren't in line with what you would expect per the values set in `scaleUpThreshold` and `scaleUpFactor` that ought to be creating new pods.

#### controller logs

in this case i kicked off a workflow with 2 jobs. (i've tried this using a range of `syncPeriod` values from `1s` to `1m+`). logs consistently showed the following output, despite one of the jobs remaining queued, even once the first job completed.

```
2021-04-21T16:54:33.911Z        DEBUG   actions-runner-controller.horizontalrunnerautoscaler    Calculated desired replicas of 1      {"horizontalrunnerautoscaler": "github-actions/ghe-runner-deployment-autoscaler", "suggested": 0, "reserved": 0, "min": 1, "cached": 0, "max": 20}
```

in instances when a job that's part of a larger workflow is queued **and** no scale up occurs, those queued jobs seem to get lost in some sort of purgatory and remain indefinitely queued until the workflow is canceled by the end user. 

here's the outcome of a few other configuration values in the context of the 1 workflow, 4 jobs example:

#### scale up configs test 1
_actions_
- 1 workflow with 4 jobs

_configs_
```
  minReplicas: 1
  maxReplicas: 20
  metrics:
  - type: PercentageRunnersBusy
    scaleUpThreshold: '0.45'    # The percentage of busy runners at which the number of desired runners are re-evaluated to scale up
    scaleDownThreshold: '0.3'   # The percentage of busy runners at which the number of desired runners are re-evaluated to scale down
    scaleUpFactor: '2.0'        # The scale up multiplier factor applied to desired count
    scaleDownFactor: '0.5'      # The scale down multiplier factor applied to desired count
```
_outcome_
- 1 job ran successfully
- 2 pods were created, though no job ever ran on the 2nd pod
- 3 jobs remained queued

#### scale up configs test 2
_actions_
- 1 workflow with 4 jobs

_configs_
```
  minReplicas: 2
  maxReplicas: 20
  metrics:
  - type: PercentageRunnersBusy
    scaleUpThreshold: '0.5'    # The percentage of busy runners at which the number of desired runners are re-evaluated to scale up
    scaleDownThreshold: '0.3'   # The percentage of busy runners at which the number of desired runners are re-evaluated to scale down
    scaleUpFactor: '3.0'        # The scale up multiplier factor applied to desired count
    scaleDownFactor: '0.5'      # The scale down multiplier factor applied to desired count
```

_outcome_
- 2 jobs ran successfully
- 6 pods were created
- 2 remaining jobs were indefinitely queued, none of the newly created pods registered there was a queue, all shared identical logs in the `runner` container (below)
- once the 2 jobs that _did_ run completed and the 2 queued jobs were still not picked up and subsequently cancelled on the github side, the pods scaled _down_ to 3
- on the github side, those 3 workers (and only those 3 workers) showed as 'idle' self-hosted workers (expected behavior)

runner container logs >>
```
Starting Runner listener with startup type: service
Started listener process
Started running service

√ Connected to GitHub

2021-04-21 17:36:32Z: Listening for Jobs
```

**further investigation notes**

prior to the above tests having run, one discrepancy i did find while investigating this was there were a number of 'offline' workers displaying in the github UI that i hadn't caught. i suspect these were left over from a prior update where i ran into an issue that's been documented in a few other posts where a seemingly infinite scale-up occurred. none of the 'ghost' workers showing in the github UI were to be found when checking kubernetes.

despite those 'offline' workers on the github side, the controller logs never reflected that number of workers, so you'd think these purgatory workers wouldn't be a factor in autoscaling in that respect, but perhaps an unrelated side-effect of an upgrade. in the instance above where the logs calculated a desired replica of 1, there were 63 'offline' workers i found at the enterprise level.

(i confirmed once i _deleted_ the 63 offline workers the behavior re: autoscaling and queued jobs was the same. on subsequent scale downs, the terminated pods were being removed as available self-hosted runners on the github side as well, so manually deleting those on the github side seemed to resolve that particular problem)

**summary of what i've tried so far**

- clean installs of `v0.18.2` with the latest `helm` charts (and manually deleting/re-creating the new CRDs), i've also tried with the `canary` image on the controller as noted above
- the manual delete/update `finalizer` approach per https://github.com/summerwind/actions-runner-controller/issues/418#issuecomment-815037707
- adjusting the scale up values by orders of magnitude
- confirmed this is an issue whether a single runner deployment with a label has been applied or two deployments with different labels. the expected behavior is there in that it only affects the deployment with the relevant label, so it doesn't seem to be related to the total number of live pods as i initially thought it could be some sort of `selector` problem.

---

my best guess right now is that there's some sort of discrepancy between how workflows and the jobs that are part of that larger workflow are registering with the controller. i don't know if it could relate to how these webhooks are treated when the runner is set at the enterprise level versus at the organization and/or repository level?

(an additional question i had on that note is whether the github 'webhook' feature can even be enabled on an enterprise level, but perhaps that warrants a separate issue)

thank you for reading this far - please let me know if you need any further info from me!! below are some more k8s specs for reference
<details>
<summary>`RunnerDeployment` spec</summary>

```
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: ghe-global-runner-deployment
  namespace: github-actions
spec:
  template:
    spec:
      enterprise: xxxxxx
      group: k8s
      labels:
        - basic
      image: summerwind/actions-runner:latest
      imagePullPolicy: Always
      dockerdWithinRunnerContainer: false
      resources:
        limits:
          cpu: "7.0"
          memory: "7Gi"
        requests:
          cpu: "7.0"
          memory: "7Gi"
      dockerdContainerResources:
        limits:
          cpu: "7.0"
          memory: "7Gi"
        requests:
          cpu: "7.0"
          memory: "7Gi"
      tolerations:
        - key: "node.kubernetes.io/unreachable"
          operator: "Exists"
          effect: "NoExecute"
          tolerationSeconds: 10
```

</details>

<details>
<summary>`HorizontalRunnerAutoscaler` specs</summary>

```
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: ghe-runner-deployment-autoscaler
  namespace: github-actions
spec:
  scaleDownDelaySecondsAfterScaleOut: 10
  scaleTargetRef:
    name: ghe-global-runner-deployment
  minReplicas: 2
  maxReplicas: 20
  metrics:
  - type: PercentageRunnersBusy
    scaleUpThreshold: '0.5'    # The percentage of busy runners at which the number of desired runners are re-evaluated to scale up
    scaleDownThreshold: '0.3'   # The percentage of busy runners at which the number of desired runners are re-evaluated to scale down
    scaleUpFactor: '3.0'        # The scale up multiplier factor applied to desired count
    scaleDownFactor: '0.5'      # The scale down multiplier factor applied to desired count
```

</details>

<details>
<summary>`ActionsRunnerController`</summary>

```
Name:                   actions-runner-controller
Namespace:              actions-runner-system
CreationTimestamp:      Wed, 14 Apr 2021 17:28:40 -0500
Labels:                 app.kubernetes.io/instance=actions-runner-controller
                        app.kubernetes.io/managed-by=Helm
                        app.kubernetes.io/name=actions-runner-controller
                        app.kubernetes.io/version=0.18.2
                        helm.sh/chart=actions-runner-controller-0.11.0
Annotations:            deployment.kubernetes.io/revision: 17
                        meta.helm.sh/release-name: actions-runner-controller
                        meta.helm.sh/release-namespace: actions-runner-system
Selector:               app.kubernetes.io/instance=actions-runner-controller,app.kubernetes.io/name=actions-runner-controller
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           app.kubernetes.io/instance=actions-runner-controller
                    app.kubernetes.io/name=actions-runner-controller
  Annotations:      kubectl.kubernetes.io/restartedAt: 2021-04-20T17:57:51-05:00
  Service Account:  actions-runner-controller
  Containers:
   manager:
    Image:      summerwind/actions-runner-controller:canary
    Port:       9443/TCP
    Host Port:  0/TCP
    Command:
      /manager
    Args:
      --metrics-addr=127.0.0.1:8080
      --enable-leader-election
      --sync-period=10s
      --docker-image=docker:dind
    Environment:
      GITHUB_TOKEN:           <set to the key 'github_token' in secret 'controller-manager'>  Optional: false
      GITHUB_ENTERPRISE_URL:  [redacted]
    Mounts:
      /etc/actions-runner-controller from secret (ro)
      /tmp from tmp (rw)
      /tmp/k8s-webhook-server/serving-certs from cert (ro)
   kube-rbac-proxy:
    Image:      quay.io/brancz/kube-rbac-proxy:v0.8.0
    Port:       8443/TCP
    Host Port:  0/TCP
    Args:
      --secure-listen-address=0.0.0.0:8443
      --upstream=http://127.0.0.1:8080/
      --logtostderr=true
      --v=10
    Environment:  <none>
    Mounts:       <none>
  Volumes:
   secret:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  controller-manager
    Optional:    false
   cert:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  webhook-server-cert
    Optional:    false
   tmp:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  <none>
NewReplicaSet:   actions-runner-controller-7d9467c88 (1/1 replicas created)
Events:          <none>
```

</details>

**Comment by callum-tait-pbx** [comment]
Great detailed post!

I actually recently updated the docs on the enteprise support https://github.com/summerwind/actions-runner-controller/pull/460.

> A RunnerDeployment (excluding enterprise runners)

Unfortunately as stated in the README ^, autoscaling runners is not supported at the enterprise level (see PR for more details). I'm not sure if it isn't there because it just isn't possible with the APIs that currently exist or it isn't there because it just hasn't been implemented. PRs are welcomed from the community for enhancing and maintaining features targeting the enterprise specific feature set though! 

I know this is a bit cheeky as I have said the thing you want isn't supported atm but I suspect that the `Runner` kind also works at the Enteprise level and that it is only autoscaling that is currently missing feature wise. Could you try deploying a `Runner` kind at the enterprise level and seeing if it can be used by an org? I'd like to update the docs to reflect this if so.

https://github.com/summerwind/actions-runner-controller/pull/290 Enterprise runner support was added here by the community, @zetaab the author of this feature confirmed in https://github.com/summerwind/actions-runner-controller/pull/460 that enterprise runners do not support autoscaling, perhaps he can tell us if autoscaling isn't there because of a technical limitation or if he just didn't need the feature and so didn't implement support.

**Comment by zetaab** [comment]
the problem is that in Enterprise level there is no way to setup webhooks that will inform controller about the need of scaling up. 
![Screenshot 2021-04-22 at 17 52 33](https://user-images.githubusercontent.com/22482503/115735964-8e6d9b80-a393-11eb-8863-c096e1b608f9.png)

these are the only supported event types in enterprise level - and that is not enough. Like in code https://github.com/summerwind/actions-runner-controller/blob/master/controllers/horizontal_runner_autoscaler_webhook.go#L144-L197 we can see we need events that are not supported by Github Enterprise.

**Comment by callum-tait-pbx** [comment]
This is just in the case of if you are using the optional [webhook](https://github.com/summerwind/actions-runner-controller#faster-autoscaling-with-github-webhook) based scaling though right? What about for the pull based metrics like `PercentageRunnersBusy`? In this case are you not just doing the following at every sync period:

1. https://docs.github.com/en/rest/reference/enterprise-admin#list-self-hosted-runner-groups-for-an-enterprise getting a list of the runner groups
2. https://docs.github.com/en/rest/reference/enterprise-admin#list-self-hosted-runners-in-a-group-for-an-enterprise get a list of the runners in each group and their status
3. scaling up or down based on the configured thresholds and current runner states?

**Comment by zetaab** [comment]
that could work

**Comment by kathleenfrench** [comment]
@callum-tait-pbx thanks for your quick response! i admittedly missed that update to the readme on this issue, which is my bad.

> Could you try deploying a Runner kind at the enterprise level and seeing if it can be used by a repo in a org? I'd like to update the docs to reflect this if so.

i'll give this a try today and let you know!

> enterprise runners do not support autoscaling, perhaps he can tell us if autoscaling isn't there because of a technical limitation or if he just didn't need the feature and so didn't implement support.

and then @zetaab confirmed this is the case specifically on the webhook autoscaler metric (as there is not an enterprise-level webhook for events).

in my particular case, I have been trying to autoscale using the `PercentageRunnersBusy` metric which for all intents and purposes appears to function exactly the same as it would were it scoped to an organization or repository. i ran a bunch of other tests yesterday to try to isolate the behavior a bit more, and can confirm that when running `RunnerDeployment`s at the enterprise level and - at least what I'm seeing - is that only those workers that exist at the time a workflow and/or job is triggered will ever be able to run those queued workflows/jobs.

what isn't clear to me still is why the autoscaler _does_ seem to ultimately work (though not within the `syncPeriod`) on the controller side (per that scale up from 2 -> 6 example) given the response from the Github API accurately shows that 2 runners are busy. then, once they _do_ scale up, the problem becomes why those new deployments aren't able to detect queued jobs, despite the logs indicating they're 'listening' for them. so the problem is twofold:

1. autoscaling happens, just not at the right time
2. upon autoscaling, new workers are not able to detect queued workflows/jobs - only those worker deployments that existed at the time they were triggered can/do (though they don't always - some jobs simply hang indefinitely even once an available worker deployment becomes available).

i went down a bit of a rabbit hole on this on the [`actions/runner`](https://github.com/actions/runner) side of things. the enterprise docs state that [here](https://docs.github.com/en/actions/hosting-your-own-runners/using-self-hosted-runners-in-a-workflow#routing-precedence-for-self-hosted-runners) that:

> When routing a job to a self-hosted runner, GitHub looks for a runner that matches the job's runs-on labels:
```
- GitHub first searches for a runner at the repository level, then at the organization level.
- The job is then sent to the first matching runner that is online and idle.
- If all matching online runners are busy, the job will queue at the level with the highest number of matching online runners.
- If all matching runners are offline, the job will queue at the level with the highest number of matching offline runners.
- If there are no matching runners at any level, the job will fail.
- If the job remains queued for more than 24 hours, the job will fail.
```

i'm admittedly not proficient in C# so it's taking me a bit to mine through that codebase, but thus far i haven't been able to identify what the disconnect might be between the repository level search -> org-level search -> enterprise level search pipeline wherein queued jobs are getting lost unless a deployment that existed at the time in which they were created exist.

there are a few things in the `/runner/_diag` logs i found that were a bit strange that seem to recur in a number of the existing `actions/runner` issues, but it remains unclear if they're in any way related to the problems seen here. despite the fact the installation i have doesn't use azure at all, it seems as though there's an authentication error getting thrown each time workers initialize that points to some sort of azureop service. 

> [2021-04-22 02:33:38Z INFO RunnerServer] Establish connection with 60 seconds timeout.
[2021-04-22 02:33:38Z INFO GitHubActionsService] Starting operation Location.GetConnectionData
[2021-04-22 02:33:38Z INFO GitHubActionsService] Finished operation Location.GetConnectionData
[2021-04-22 02:33:38Z INFO GitHubActionsService] Finished operation Location.GetConnectionData
[2021-04-22 02:33:38Z INFO GitHubActionsService] Finished operation Location.GetConnectionData
[2021-04-22 02:33:39Z INFO MessageListener] VssConnection created
[2021-04-22 02:33:39Z INFO Terminal] WRITE LINE:
[2021-04-22 02:33:39Z INFO Terminal] WRITE LINE:
[2021-04-22 02:33:39Z WARN GitHubActionsService] Authentication failed with status code 401.
Date: Thu, 22 Apr 2021 02:33:39 GMT
Server: Kestrel
Transfer-Encoding: chunked
WWW-Authenticate: Bearer
x-tfs-processid: c6cfb4d1-ddfd-486a-a15e-9cb06bc5a0c2
activityid: 82360375-d988-4a0b-8128-d4aab9891dc1
x-tfs-session: 7ab45133-057d-4145-bf79-e9a7747cf7af
x-vss-e2eid: 6f74991a-b075-4d33-bc8d-107ef8f7bde6
x-vss-senderdeploymentid: 2d75ebf9-dfb6-0ce6-d1b9-7b225cc85ddf
x-tfs-serviceerror: The+user+%27System%3aPublicAccess%3baaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa%27+is+not+authorized+to+access+this+resource.
x-vss-s2stargetservice: 0000005A-0000-8888-8000-000000000000/visualstudio.com

https://github.com/actions/runner/issues/679 documents this issue and the response from github seems to be that a fix was deployed (in September) that ought to have redressed it

> We’re aware of the current issue where runners added after queuing a run will not pick up the jobs if the added runner is not the latest version. We’re rolling out the fix and will update the issue again once it’s deployed everywhere. In meantime, you can unblock yourselves by adding the latest runner every time.

so i forked this repo and built docker images with the latest runner release, which as of writing is `2.278.0`. after trying to run workflows with that, i can confirm the issue persists.

it seems as though the scale-up issue is coming from the controller here and the failure to assign jobs to new workers is coming from the `actions/runner` side. a low syncPeriod like `10s` ought to be detecting the need for new workers based on the scale up thresholds set in my initial examples, so perhaps it's related to something with how those listed runners are being cached? where i'm still in the dark is why the new workers that eventually come up aren't seeing those workers, and i'm leaning towards that being a bug on the runner side per the github rep's comment above (as this was a known issue in previous iterations).

the only other things i found in some `/runner/_diag` logs of note is the following log error on a new worker deployment. this wasn't happening consistently, so i can't say for sure what action, if any, it correlates to:

<details>
<summary>error log > click me</summary>

```
[2021-04-22 02:50:14Z INFO Terminal] WRITE LINE:
[2021-04-22 02:50:14Z INFO ConfigurationManager] Test Connection complete.
[2021-04-22 02:50:14Z INFO HostContext] Well known directory 'Bin': '/runner/bin'
[2021-04-22 02:50:14Z INFO HostContext] Well known directory 'Root': '/runner'
[2021-04-22 02:50:14Z INFO HostContext] Well known config file 'RSACredentials': '/runner/.credentials_rsaparams'
[2021-04-22 02:50:14Z INFO RSAFileKeyManager] Creating new RSA key using 2048-bit key length
[2021-04-22 02:50:14Z INFO RSAFileKeyManager] Successfully saved RSA key parameters to file /runner/.credentials_rsaparams
[2021-04-22 02:50:14Z INFO RSAFileKeyManager] Which: 'chmod'
[2021-04-22 02:50:14Z INFO RSAFileKeyManager] Location: '/usr/bin/chmod'
[2021-04-22 02:50:14Z INFO HostContext] Well known directory 'Bin': '/runner/bin'
[2021-04-22 02:50:14Z INFO HostContext] Well known directory 'Root': '/runner'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper] Starting process:
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   File name: '/usr/bin/chmod'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   Arguments: '600 /runner/.credentials_rsaparams'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   Working directory: '/runner'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   Require exit code zero: 'False'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   Encoding web name:  ; code page: ''
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   Force kill process on cancellation: 'False'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   Redirected STDIN: 'False'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   Persist current code page: 'False'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   Keep redirected STDIN open: 'False'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper]   High priority process: 'False'
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper] Failed to update oom_score_adj for PID: 63.
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper] System.UnauthorizedAccessException: Access to the path '/proc/63/oom_score_adj' is denied.
 ---> System.IO.IOException: Permission denied
   --- End of inner exception stack trace ---
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode)
   at System.IO.FileStream.OpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.StreamWriter.ValidateArgsAndOpenPath(String path, Boolean append, Encoding encoding, Int32 bufferSize)
   at System.IO.StreamWriter..ctor(String path)
   at System.IO.File.WriteAllText(String path, String contents)
   at GitHub.Runner.Sdk.ProcessInvoker.WriteProcessOomScoreAdj(Int32 processId, Int32 oomScoreAdj)
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper] Process started with process id 63, waiting for process exit.
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper] STDOUT/STDERR stream read finished.
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper] STDOUT/STDERR stream read finished.
[2021-04-22 02:50:14Z INFO ProcessInvokerWrapper] Finished process 63 with exit code 0, and elapsed time 00:00:00.0160661.
[2021-04-22 02:50:14Z INFO RSAFileKeyManager] Successfully set permissions for RSA key parameters file /runner/.credentials_rsaparams
[2021-04-22 02:50:14Z INFO CommandSettings] Arg 'runnergroup': 'k8s'
[2021-04-22 02:50:14Z INFO CommandSettings] Remove runnergroup from Arg dictionary.
[2021-04-22 02:50:14Z INFO Terminal] WRITE LINE:
[2021-04-22 02:50:14Z INFO ConfigurationManager] Found a self-hosted runner group with id {1} and name {2} 2, k8s
[2021-04-22 02:50:14Z INFO CommandSettings] Arg 'name': 'ghe-global-runner-deployment-p9f8d-vbwfl'
[2021-04-22 02:50:14Z INFO CommandSettings] Remove name from Arg dictionary.
```

</details>

i'll update with more as i find it, but if you have any guesses as to the two core problems re: autoscaling and failure to identify queued workflows/jobs i'm all ears 😀 

thanks again for your help!

**Comment by kathleenfrench** [comment]
ah, you know what i think i may have found the solution to the autoscaling issue [here](https://github.com/summerwind/actions-runner-controller/blob/374105c1f30ce8cc1e54079b432027ff63521463/main.go#L159)

```
	horizontalRunnerAutoscaler := &controllers.HorizontalRunnerAutoscalerReconciler{
		Client:        mgr.GetClient(),
		Log:           log.WithName("horizontalrunnerautoscaler"),
		Scheme:        mgr.GetScheme(),
		GitHubClient:  ghClient,
		CacheDuration: syncPeriod - 10*time.Second,
	}
```

**CacheDuration: syncPeriod - 10*time.Second** - this meant that my tests trying this on a really low sync interval (which were between 2-10s in the majority if not all of the autoscaling tests as luck would have it) are falling back to a default 10 minute cache [here](https://github.com/summerwind/actions-runner-controller/blob/374105c1f30ce8cc1e54079b432027ff63521463/controllers/horizontalrunnerautoscaler_controller.go#L137)

```
		var cacheDuration time.Duration

		if r.CacheDuration > 0 {
			cacheDuration = r.CacheDuration
		} else {
			cacheDuration = 10 * time.Minute
		}
```

building an image to test it out now, but willing to bet - at least on the autoscaling front - that's why a scale-up wasn't happening. i'll confirm whether or not that's the case in a bit.

**Comment by kathleenfrench** [comment]
ok, the good news is that **did** fix autoscaling to go in and reduce the cache duration - now scale-ups and scale-downs happen when you'd expect if you're trying to detect autoscaling on a much shorter interval (just make sure you don't have a low `scaleDownDelaySecondsAfterScaleOut` value or you'll get stuck in a recreation loop)

the bad news is the failure to detect queued workflows/jobs remains. jobs will only run on runner pods that existed at the time the workflow/job triggers - so, i'll keep digging on that front 

it does seem that on the pods that actually run the jobs a `Worker_` log gets created in the `/runner/_diag`  directory (along with the `Runner_` logs), whereas in new pods only `Runner_` logs are ever created?

**Comment by callum-tait-pbx** [comment]
Great to hear! Hopefully we can get the assignment piece working and then we have autoscaling everywhere! 💪 

> matching is what sticks out to me - matching as in matching the deployment name in some job agent cache on the runner side perhaps?

It's supposed to be just matching in terms of labels on the workflow via the `runs-on` key and the labels assigned to the runners. If there is more to it then GitHub don't talk about it! It might be worth raising an issue with https://github.com/actions/runner to see if there is more to the logic than that and if there is a way of seeing what GitHub is seeing in terms of avaliable runners for the matching set of labels?

**Comment by callum-tait-pbx** [comment]
https://github.com/actions/runner/issues/1059 interesting, it looks like the problem lies with the github routing service on github's end.

For the moment then it looks like enterprise level autoscaling isn't possible until GitHub improve their queueing / routing service, lame. If they do end up improving their service the code as it stands may already be able to take advantage of feature.

@kathleenfrench if you could try deploying a `Runner` kind and see if it can be consumed by repositories in organisations that would be great! We can then update the docs with all this great information you've managed to discover!

In the meantime if you raise something on https://github.community/c/code-to-cloud/github-actions/41 as advised by github we can link it to an issue here so we can keep track of it.

**Comment by kathleenfrench** [comment]
@callum-tait-pbx ran some more tests this morning, and there's some good news on the enterprise autoscaling front (though it does beg the utility of an enterprise-level `RunnerDeployment` as Github's queuing/routing currently stands)

I created a `RunnerDeployment` mirroring the configurations of the enterprise `RunnerDeployment` (i.e. replacing `enterprise: [value]` with `organization: [value]`) and upon triggering the same workflows as I did in the previous examples, the `organization` runners were the only ones triggered to scale-up and _unlike the enterprise runners, they **do** pick up those queued jobs_. 

I think this pretty much confirms the problem is on Github's side for how queuing/routing is handled (and the response to the issue I opened in the `actions/runner` repo you link to seems to indicate via their response that to some extent this is a known problem).

I will raise something in the github community forum on the subject and link to it here in a follow-up. i wonder if the lack of job-related events reported (or at least accessible) via the enterprise level webhook as @zetaab has already pointed out (which currently only exposes user management tasks) is part of the underlying problem. 

for those interested, with one enterprise level deployment along with other `organization` level deployments, you never see a scale-up happen on the `enterprise` `RunnerDeployment`. if you were to trigger a job from some other organization or deployment not covered by `RunnerDeployment` entities, those will be picked up by the enterprise runner with the caveat that they'll only run if you have enough runners already up at the time the workflows kick off. 

so, if you opt to use enterprise level deployments, you need to set a `minReplicas` value that you're confident will provide coverage for any and all workflows that could be triggered at any given time. autoscaling will function as expected on the controller side in those instances, and once those runners come up any subsequent jobs triggered in the interim prior to a scale down _will_ be picked up by those runners **but** if a workflow/job is triggered and there is no enterprise-level runner available it will be lost to a queue purgatory. if you really want to keep them at an enterprise level, then, you want to use a large scale up factor and prevent immediate scale downs so new workflows get picked up. this sort of undermines the larger aim of autoscaling, but if perhaps you don't want to create an org or repository scoped runner to cover some rare job, enterprise runners will work, the minimum just needs to reflect how many you'd need at run time, for lack of better phrasing. 

if you have `organization` level `RunnerDeployment`s the above problem is not the case - autoscaling will work as expected _and_ newly scaled runners will receive queued jobs.

i'm running all of these in the same namespace without any problems, but as this is on GHES we're able to disable rate-limiting for this, as I know the multiple namespaces solution has been necessary in other instances where that's a limiting factor.

once Github fixes the problems with the enterprise queue/routing, though, i'm confident the `actions-runner-controller` will work as it's currently configured. 

as for the `CacheDuration` change I noted above, shaving off those extra seconds probably wouldn't be utilized by a great deal of end-users, but perhaps offering a way to set that subtracted value instead of the 10s default would be worthwhile? or at the very least, ensuring that if the `syncPeriod` is lower than the default `CacheDuration` that it won't return a negative value (which, as it currently stands, results in falling back to a 10m cache interval).

**Comment by callum-tait-pbx** [comment]
@kathleenfrench  Did you get a chance to deploy a `Runner` kind at the enterprise level?

**Comment by kathleenfrench** [comment]
@callum-tait-pbx I didn't realize the `Runner` could be set to enterprise, sorry I didn't realize that's what you meant! 

just tried that and if I add a `Runner` enterprise-level deployment it appears to suffer from the same problem in that only those workflows/jobs immediately allocated to a runner get picked up, but they _do_ work in that case. 

is there a way to ref a `Runner` to some sort of horizontal autoscaling, though? my understanding was that was limited to `RunnerDeployment` types (and at least in an attempt to try it on my end by target ref'ing a runner there wasn't any scale up at all) which I'm pretty sure is expected, but want to make sure I'm not misreading.

**Comment by callum-tait-pbx** [comment]
> @callum-tait-pbx I didn't catch that the Runner could be set to enterprise, sorry I didn't realize that's what you meant!

no problem!

> just tried that and if I add a Runner (enterprise-level) it appears to suffer from the same problem in that only those workflows/jobs immediately allocated to a runner get picked up, but they do work in that case./

Sounds like it works! The `Runner` kind is for deploying a single runner rather than sets. It doesn't support being scaled by a `horizontalrunnerautoscaler`. I'm not sure how much use it is at an enterprise level really tbh but I was pretty sure it worked I just needed someone to test it for me :D. With that confirmed the docs can read and flow better.

**Comment by kathleenfrench** [comment]
one other relevant observation is that if there are _no_ organization-level `RunnerDeployment` or `Runner` pods up at the time a workflow is triggered, then a subsequent creation of those also fail to detect there are queued jobs. so while the org-level runners _will_ successfully find queued jobs this is _only_ the case if there is one or more runner of that type live at the time the workflow began

**Comment by callum-tait-pbx** [comment]
> so while the org-level runners will successfully find queued jobs this is only the case if there is one or more runner of that type live at the time the workflow began

If I am understanding you correctly I think this is known github behaviour. If there are no matching runners registered in any state (offline, idle, running) then a workflow will immediately fail.

**Comment by kathleenfrench** [comment]
@callum-tait-pbx this is while there is a `RunnerDeployment` already up for the enterprise level, though - I wanted to see if perhaps the failure for the enterprise deployments to detect queued jobs was limited just to deployments of that type, so I tried first deleting the org-level deployment that was running alongside the enterprise-level deployment, triggering a workflow with 4 jobs while 1 enterprise deployment was live (which, as we now know, always fails to detect the queue of 3 other jobs), then I added an org-level deployment and autoscaler to the mix. 

since the org `RunnerDeployment` _does_ actually run those queued jobs _after_ a scale-up, I wanted to see whether they would detect those queued jobs when they're deployed after the workflow is initially triggered (versus in a scenario where some of those initial jobs are captured by that org deployment). 

> If there are no matching runners registered in any state (offline, idle, running) then a workflow will immediately fail.

in this case no jobs fail because there is already an enterprise-level deployment online with compatible labels, but adding new runners with compatible labels to account for those queued jobs which - in other instances successfully detect and run them assuming they're live at the time the workflow begins - are subject to the same queue purgatory problem.

**Comment by kathleenfrench** [comment]
Here's what I've posted in the github community forum 👍🏻  -> https://github.community/t/bug-self-hosted-runners-at-the-enterprise-level-fail-to-detect-queued-jobs/176348

**Comment by callum-tait-pbx** [comment]
> it seems like there must be some sort of message queue retry or caching problem on the job router side of things, which is odd because github's own API accurately reflects the state of available, online runners while these jobs remain stuck in the queue.

It's a bit of a weird one isn't it!

> as for the CacheDuration change I noted above, shaving off those extra seconds probably wouldn't be utilized by a great deal of end-users, but perhaps offering a way to set that subtracted value instead of the 10s default would be worthwhile? or at the very least, ensuring that if the syncPeriod is lower than the default CacheDuration that it won't return a negative value (which, as it currently stands, results in falling back to a 10m cache interval).

Potentially we could just remove the subtraction entirely and just check the the value provided is >= 1s and just default to the fail safe 10m figure if the input is bad or someone tries to do a sub-second sync period. The solution has improve greatly since its first inception, originally only the pull based `TotalNumberOfQueuedAndInProgressWorkflowRuns` metric was the only means for scaling, this metric requires a lot of API calls to maintain and so it wasn't really possible to run a very low sync period. Since then the `PercentageRunnersBusy` metric has been added which uses far fewer API calls to maintain as well as a webhook option which doesn't rely on lots of API calls to begin with. We've also added the ability to run multiple controllers since then too to allow very large scale setups without getting rate limited.

The default 10 minutes for a sync period was chosen just to stop someone rate limiting themselves out of the box, there isn't anything particular about why it's 10 minutes beyond that reason.

Alternatively if we feel 1s is too low and you are extremely likely to rate limit yourself then we could keep the check at the 10s threshold (or maybe only lower it to 5s) and if the sync period provided is below that threshold just default to the lowest accepted value instead of the fail safe and spit out some log messages so it is debugable. The logic being the end user as actively chosen a short sync period then we are assuming they understand they may get rate limited.

cc @mumoshu what are your thoughts?

**Comment by callum-tait-pbx** [comment]
For posterity's sake, I raised this with our Enterprise support

![image](https://user-images.githubusercontent.com/52924845/116216565-ab61f000-a740-11eb-9b26-781ec4b5d956.png)

**Comment by mumoshu** [comment]
@kathleenfrench @callum-tait-pbx Hey! Sorry for taking so long to respond to this great thread.

I just read https://github.community/t/bug-self-hosted-runners-at-the-enterprise-level-fail-to-detect-queued-jobs/176348 and GitHub is going to fix the issue on their end. Good.

Until the fix lands, the best we can do in terms of autoscaling right now would be to do update HRA to add [CapacityReservation](https://github.com/actions-runner-controller/actions-runner-controller/blob/7e766282aadc1b13f81515609151470c8eb56b70/api/v1alpha1/horizontalrunnerautoscaler_types.go#L56) to add runners for a specific period of time. If you're workflow runs/jobs are being run on schedule/timings are predictable, this workaround would work.

Another possible solution would be "scheduled scaling" feature we're discussing in #484. It would allow you to have more runners in business hours and fewer runners in other hours. The scaling isn't as dynamic as it would be with PercentageRunnersBusy though.

Regarding `CacheDuration`, I'm open to making it more customizable.

What if we had another flag `--github-api-cache-duration` to customize it, so that you can freely choose whatever `--sync-period` and `--github-api-cache-duration` values you want?

**Comment by callum-tait-pbx** [comment]
> We are planning to ship a fix for this behavior in the next few weeks. Given that it is job assignment logic, though, it might take us some time to test and verify the behavior as we roll it out (breaking job assignment would be pretty bad :smile:.

> Until the fix lands, the best we can do in terms of autoscaling right now would be to do update HRA to add CapacityReservation to add runners for a specific period of time. If you're workflow runs/jobs are being run on schedule/timings are predictable, this workaround would work.

Given the time horizon here from GitHub is a few weeks for the fix to be shipped and then I think realistically it could take up to a month for it to be fully tested on their end before general rollout I personally wouldn't want to start introducing code that is only temporary fix unless it added broader long term value to the project, 

> i've configured the controller on the enterprise level, as we have many organizations to support

Additionally, correct me if I'm wrong @kathleenfrench but her main driver for wanting enterprise scaling is because doing it at the organisation level is not realistic in her environment due to the number of organisations that she needs to support. I don't think more introducing more flex around how the things scale would do much for her as an end user.

> Regarding CacheDuration, I'm open to making it more customizable.

I think this is worth doing with the current value as the default. Considering Github Enterprise Server lets the adminstrator control the rate limiting behaviour and the solution supports that environment it would be handy to be able to bring that value right down. Once Github patch their end beign able to lower this value would still be very beneficial in a Github Enterprise Server environment.

**Comment by kathleenfrench** [comment]
@callum-tait-pbx 
> Additionally, correct me if I'm wrong @kathleenfrench but her main driver for wanting enterprise scaling is because doing it at the organisation level is not realistic in her environment due to the number of organisations that she needs to support. I don't think introducing more flex around how the things scale would do much for her as an end user.

the end goal is to definitely have these scale at the enterprise level to avoid needing organization-specific deployments where they otherwise wouldn't be required. in the absence of an imminent fix from github, i think @mumoshu's idea is really great, but if they're saying they're actively working on a fix for this then i'm not sure if the work required to support that here would be warranted.

i wonder if part of github's work on improving their job queue/autoscaling at the enterprise level will include expanding the scope of enterprise-level webhooks as they pertain to running actions. i suspect that's not part of the near-future roadmap, though, as i'd speculate it's a larger undertaking given the necessary changes to the existing API. 

i anticipate, then, that `PercentageRunnersBusy` will continue to be the most useful autoscaling metric for enterprise-level workers for the foreseeable future, and given that what _could_ be useful is some sort of feature that takes from @mumoshu's idea to include `CapacityReservation` hours that would support different scale up/down factors according to the hours where there is more traffic. 

as every job spawns its own runner pods, and a single workflow can include so many jobs, when you're running at the enterprise level the scale up factor may have to be higher than it would otherwise be at a repository/org-level, but you wouldn't necessarily want that to be the behavior for jobs running at non-peak traffic hours/non-business hours. that might be a bit too in the weeds as a feature for enterprise users, but it's an idea.

**Comment by callum-tait-pbx** [comment]
@kathleenfrench thanks for the great response as usual.

> i wonder if part of github's work on improving their job queue/autoscaling at the enterprise level will include expanding the scope of enterprise-level webhooks as they pertain to running actions. i suspect that's not part of the near-future roadmap, though, as i'd speculate it's a larger undertaking given the necessary changes to the existing API.

Yes I took that to mean they are patching their job assignment logic rather than expanding out the events at the Enterprise level. I assume the latter is much more involved and has implications around their infrastructure and so would not be part of the fix they are applying. Hopefully they eventually expand out the events though in future!

> as every job spawns its own runner pods, and a single workflow can include so many jobs, when you're running at the enterprise level the scale up factor may have to be higher than it would otherwise be at a repository/org-level, but you wouldn't necessarily want that to be the behavior for jobs running at non-peak traffic hours/non-business hours. that might be a bit too in the weeds as a feature for enterprise users, but it's an idea.

Have you seen the new scheduling feature @mumoshu has managed to put together? Would this not cover your need to run a larger amount of min replicas during peak hours and then scale back duing non peak/non-business hours?

**Comment by hross** [comment]
Rather than let you all speculate in this issue, I'll just go ahead and put some rough plans in here. We plan to ship most of this by the end of the summer and we plan to also ship it in GHES after we test it on dotcom.

#### Stuff that should land in June-ish:

* Fix job assignment to allow it at multiple levels
* Change job failure timeouts so that runs won't fail if they don't find an existing runner (online or offline) with a label that matches. This allows you to do on-demand spin up rather than register a "fake" runner.
* Enable the [ephemeral](https://github.com/actions/runner/pull/660) flag to truly guarantee run once jobs. We can land this after the above changes.

#### Will probably take longer due to figuring out API contracts and performance testing:

* Add webhooks similar to check runs -- queue, running, completed -- (except workflow specific) that flow through the org, repo and enterprise levels. This would also include attaching labels to the events and adding a durable job ID so you can reconcile missed events or issues with your scaling solution. Possibly runner group information (but maybe add that later). **I think this is exactly what you're talking about above 😄**

I"m happy to work with you all to test these out before they are "officially" available so you can update the solution. We think this is a lot of good stuff to enable auto-scaling runners. If you have feedback, want to get in touch about implementation or think there are more things we should do in this area you can email me using my handle at github.

**Comment by toast-gear** [comment]
@hross really excited for when these changes to land!

I just wanted to highlight https://github.com/actions-runner-controller/actions-runner-controller/issues/642 to you as we've had a number of people request the ability to scale their runner counts up based on label from webhooks. At the moment the payloads don't include the information we would need to consider doing this in the project.

**Comment by sammcj** [comment]
Trying to debug autoscaling Actions runners and I came across this thread.

I see that autoscaling runners aren't working at an _enterprise_ level, but are they working at an **_organisation_** level?

e.g. runners added at organisational settings -> actions -> runners

**Comment by hross** [comment]
👋 We were able to ship the assignment changes last week, so you should now see jobs get picked up at all levels even if you register a runner after a job was scheduled (including organization level).

Make sure you check your runner group [access permissions](https://docs.github.com/en/actions/hosting-your-own-runners/managing-access-to-self-hosted-runners-using-groups#changing-the-access-policy-of-a-self-hosted-runner-group) when you debug this.

We are working on the webhooks as we speak and still expect to ship ephemeral soon (September time frame).

**Comment by mumoshu** [comment]
@hross Thanks a lot for the update!

Apparently, it seems like we no longer need [registration-only runners](https://github.com/actions-runner-controller/actions-runner-controller#autoscaling-tofrom-0).

From my limited testing, I can see that GitHub can now start waiting until any runner to pick the enqueued job, even when there're no runners at all when the job was queued.

**Comment by mumoshu** [comment]
Probably @kathleenfrench's original issue of enterprise runners can't be auto-scaled at all is now resolved, too? @kathleenfrench Would you mind giving it a try when you have time?

**Comment by toast-gear** [comment]
@hross this is great to hear, I'm sure it will be updated once you are confident with the change and it's considered validated but I thought I'd highlight https://docs.github.com/en/actions/hosting-your-own-runners/using-self-hosted-runners-in-a-workflow#routing-precedence-for-self-hosted-runners needs updating to reflect the new assignment logic

**Comment by hross** [comment]
@toast-gear thanks I fixed the typo! Will also take a look at the docs to see what updates we need to make.

**Comment by hross** [comment]
just wanted to post here to let you all know we just officially launched ephemeral: https://github.blog/changelog/2021-09-20-github-actions-ephemeral-self-hosted-runners-new-webhooks-for-auto-scaling/

**Comment by toast-gear** [comment]
Thanks @hross, we are keeping this open for the moment as we're hoping @kathleenfrench will be able to test before we close. She might be on holiday atm or something though, if we don't get a response in a week or so we'll consider the issue resolved and close the ticket and deal with any issues that arise.

**Comment by pberthonneau** [comment]
May I ask @hross if `workflow_jobs` webhooks will be included in final GHES 3.2 or if we will have to wait 3 months for the 3.3?

**Comment by hross** [comment]
Unfortunately it did not make the cutoff for 3.2 release cycle and we will ship it in 3.3.

**Comment by toast-gear** [comment]
Now the docs are updated this can be considered done
